algorithms/kernel/objective_function/mse/mse_dense_default_batch_impl.i

```diff
@@ -229,60 +170,36 @@ inline services::Status MSEKernel<algorithmFPType, method, cpu>::compute(Numeric
-                                    PRAGMA_IVDEP
-                                    PRAGMA_VECTOR_ALWAYS
-                                    for (size_t i = startRow; i < finishRow; i++)
-                                    {
-                                        hessianDiagonalLocal[j] += X[j * n + i] * X[j * n + i]; /*USE DOTPRODUCT or parallel computation*/
-                                    }
-                                }
-                            });
-                            tlsData.reduce([&](algorithmFPType * localHes) {
-                                PRAGMA_IVDEP
-                                PRAGMA_VECTOR_ALWAYS
-                                for (size_t j = 0; j < nTheta; j++)
-                                {
-                                    hessianDiagonalPtr[j] += localHes[j];
-                                }
-                            });
-                        }
-                        else
-                        {
-                            TlsMem<algorithmFPType, cpu, services::internal::ScalableCalloc<algorithmFPType, cpu> > tlsData(nTheta);
-                            daal::threader_for(nBlocks, nBlocks, [&](const size_t iBlock) {
-                                algorithmFPType * hessianDiagonalLocal = tlsData.local();
-                                const size_t startRow                  = iBlock * blockSize;
-                                const size_t finishRow                 = (iBlock + 1 == nBlocks ? nDataRows : (iBlock + 1) * blockSize);
+                            ReadRows<algorithmFPType, cpu> xptr(dataNT, startRow, localBlockSize);
+                            DAAL_CHECK_BLOCK_STATUS_THR(xptr);
+                            const algorithmFPType * x = const_cast<algorithmFPType *>(xptr.get());
 
-                                for (size_t i = startRow; i < finishRow; i++)
-                                {
-                                    PRAGMA_IVDEP
-                                    PRAGMA_VECTOR_ALWAYS
-                                    for (size_t j = 0; j < nTheta; j++)
-                                    {
-                                        hessianDiagonalLocal[j] += X[i * dim + j] * X[i * dim + j]; /*USE DOTPRODUCT or parallel computation*/
-                                    }
-                                }
-                            });
-                            tlsData.reduce([&](algorithmFPType * localHes) {
+                            for (size_t i = 0; i < localBlockSize; i++)
+                            {
                                 PRAGMA_IVDEP
                                 PRAGMA_VECTOR_ALWAYS
                                 for (size_t j = 0; j < nTheta; j++)
                                 {
-                                    hessianDiagonalPtr[j] += localHes[j];
+                                    hessianDiagonalLocal[j] += x[i * dim + j] * x[i * dim + j]; /*USE DOTPRODUCT or parallel computation*/
```

Comment:
Might be better to use BLAS dot product?

Reply:
yes, comment was waiting for the refactoring :)